libpromises/evalfunction.c

```diff
@@ -789,6 +785,122 @@ static FnCallResult FnCallIP2Host(ARG_UNUSED EvalContext *ctx,
 
 /*********************************************************************/
 
+static FnCallResult FnCallSysctlValue(ARG_UNUSED EvalContext *ctx,
+                                      ARG_UNUSED ARG_UNUSED const Policy *policy,
+                                      const FnCall *fp,
+                                      const Rlist *finalargs)
+{
+#ifdef __linux__
+    const bool sysctlvalue_mode = (strcmp(fp->name, "sysctlvalue") == 0);
+
+    size_t max_sysctl_data = 16 * 1024;
+    Buffer *procrootbuf = BufferNew();
+    // Assumes that FILE_SEPARATOR is /
+    BufferAppendString(procrootbuf, GetRelocatedProcdirRoot());
+    BufferAppendString(procrootbuf, "/proc/sys");
+
+    if (sysctlvalue_mode)
+    {
+        Buffer *key = BufferNewFrom(RlistScalarValue(finalargs),
+                                    strlen(RlistScalarValue(finalargs)));
+
+        Buffer *filenamebuf = BufferCopy(procrootbuf);
+        // Assumes that FILE_SEPARATOR is /
+        BufferAppendChar(filenamebuf, '/');
+        BufferSearchAndReplace(key, "\\.", "/", "gT");
+        BufferAppendString(filenamebuf, BufferData(key));
+        BufferDestroy(key);
+
+        if (IsDir(BufferData(filenamebuf)))
+        {
+            BufferDestroy(procrootbuf);
+            BufferDestroy(filenamebuf);
+            return FnFailure();
+        }
+
+        Writer *w = NULL;
+        bool truncated = false;
+        int fd = safe_open(BufferData(filenamebuf), O_RDONLY | O_TEXT);
+        if (fd >= 0)
+        {
+            w = FileReadFromFd(fd, max_sysctl_data, &truncated);
+            close(fd);
+        }
+
+        if (w == NULL)
+        {
+            Log(LOG_LEVEL_INFO, "Error while reading file '%s' (%s)",
+                BufferData(filenamebuf), GetErrorStr());
+            BufferDestroy(procrootbuf);
+            BufferDestroy(filenamebuf);
+            return FnFailure();
+        }
+
+        BufferDestroy(procrootbuf);
+        BufferDestroy(filenamebuf);
+
+        char *result = StringWriterClose(w);
+        StripTrailingNewline(result, max_sysctl_data);
+        return FnReturnNoCopy(result);
+    }
+
+    JsonElement *sysctl_data = JsonObjectCreate(10);
+
+    Buffer *procrootbuf_slash = BufferCopy(procrootbuf);
+    BufferDestroy(procrootbuf);
```

Comment:
Why make a copy when you're immediately destroying the source? Just use the original instead.